log/DESIGN.md

```diff
@@ -0,0 +1,737 @@
+
+`Logger` can be extended by adding new `LoggerOption` options
+and adding new exported fields to the `LoggerConfig` struct.
+This design is already used in Trace API for getting tracers
+and in Metrics API for getting meters.
+
+Rejected alternative:
+
+- [Passing struct as parameter to LoggerProvider.Logger](#passing-struct-as-parameter-to-loggerproviderlogger).
+
+### Logger
+
+The [`Logger` abstraction](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#logger)
+is defined as an interface:
+
+```go
+type Logger interface {
+	embedded.Logger
+	Emit(ctx context.Context, record Record)
+}
+```
+
+The specification may add new operations to `Logger`.
+The interface may have methods added without a package major version bump.
+This embeds `embedded.Logger` to help inform an API implementation
+author about this non-standard API evolution.
+This approach is already used in Trace API and Metrics API.
+
+### Logger.Emit
+
+The `Emit` method implements the [`Emit a LogRecord` operation](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#emit-a-logrecord).
+
+[`Context` associated with the `LogRecord`](https://opentelemetry.io/docs/specs/otel/context/)
+is accepted as a `context.Context` method argument.
+
+Calls to `Emit` are supposed to be on the hot path.
+Therefore, in order to reduce the number of heap allocations,
+the [`LogRecord` abstraction](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#emit-a-logrecord),
+is defined as a `Record` type:
+
+```go
+type Record struct {
+	timestamp         time.Time
+	observedTimestamp time.Time
+	severity          Severity
+	severityText      string
+	body              Value
+
+	// The fields below are for optimizing the implementation of
+	// attributes.
+	front [5]KeyValue
+	nFront int // The number of attributes in front.
+	back []KeyValue
+}
+```
+
+[`Timestamp`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-timestamp)
+is accessed using following methods:
+
+```go
+func (r *Record) Timestamp() time.Time 
+func (r *Record) SetTimestamp(t time.Time) 
+```
+
+[`ObservedTimestamp`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-observedtimestamp)
+is accessed using following methods:
+
+```go
+func (r *Record) ObservedTimestamp() time.Time 
+func (r *Record) SetObservedTimestamp(t time.Time) 
+```
+
+[`SeverityNumber`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-severitynumber)
+is accessed using following methods:
+
+```go
+func (r *Record) Severity() Severity
+func (r *Record) SetSeverity(s Severity)
+```
+
+`Severity` type is defined and constants are based on
+[Displaying Severity recommendation](https://opentelemetry.io/docs/specs/otel/logs/data-model/#displaying-severity).
+Additionally, `Severity[Level]1` constants are defined to make the API more readable and user friendly.
+
+```go
+type Severity int
+
+const (
+	SeverityTrace1 Severity = iota + 1
+	SeverityTrace2
+	SeverityTrace3
+	SeverityTrace4
```

Comment:
Exported consts like these should never be defined in terms of `iota`, because `iota` is not stable over time. As an example, consider an innocuous-looking future change like the following:

```diff
 const (
 	SeverityTrace1 Severity = iota + 1
 	SeverityTrace2
 	SeverityTrace3
 	SeverityTrace4
+	SeverityTrace5
```

This changes the actual values of all subsequent const declarations in the block. SeverityDebug was previously defined as 5, and is now defined as 6. This breaks compatibility. 

Instead, define exported consts with explicit values.

```suggestion
const (
	SeverityTrace1 Severity = 1
	SeverityTrace2 Severity = 2
	SeverityTrace3 Severity = 3
	SeverityTrace4 Severity = 4
```

(and so on)

Reply:
Thanks from your comment.

1. All exported API should have unit tests to avoid making breaking changes. Here are the tests from the prototype: https://github.com/open-telemetry/opentelemetry-go/blob/e5dc7cf2f0e8b24fbcaf26cfd8b9f5fd5d4dc819/log/severity_test.go. We also have [a step](https://github.com/open-telemetry/opentelemetry-go/blob/main/RELEASING.md#breaking-changes-validation) in releasing process to double-check if there are not breaking changes in the stable API (it would be better if it is integrated  in CI, but this is a separate issue).
2. I agree that using `iota` can be more error-prone. However, take notice that it affects GoDoc (package documentation). I find `iota` cleaner when the underlying value is "not important" (has no underlying meaning). Therefore, I think that we could use exact values (instead of `iota`) for `Severity` constants as these are defined by [the specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-severitynumber). However, I prefer `iota` for `Kind` constants. PS. Take notice that someone can still make breaking changes when `iota` i used.
3. From API design perspective, this is an implementation detail and we can decide later during implementation if we should use `iota` or not.

Reply:
SIG meeting notes:
We will change to use the values explicitly, because they are defined by the specification.

Reply:
https://github.com/open-telemetry/opentelemetry-go/pull/4809/commits/9ab454632d24f97bdf967dbc4eba7604b7e86396